### ISSUE TITLE
[Feature] Add support for Gitlab

### ISSUE THREAD

**Issue by swmcc** [apoclyps/reviews#188]
I would like a feature to flip between or include gitlab reps.

**Describe the solution you'd like**
Either a flag that switches between github and gitlab repos or a way to include both in the same screen.

Am wiling to help out if you want. Do you think this might be an interesting feature? Could possibly extend to Bitbucket?

**Comment by apoclyps** [comment]
cheers @swmcc - I'm onboard for adding both. I'm happy to take a run at adding Gitlab support and as a follow it; it might be nice for you to tackle Bitbucket if you are interested?

**Comment by apoclyps** [comment]
https://pypi.org/project/python-gitlab/ looks like a suitable candidate for interacting with the Gitlab API via python:

python-gitlab has a similar API to pygithub: https://python-gitlab.readthedocs.io/en/stable/api-usage.html

**Comment by apoclyps** [comment]
Initial spike of python-gitlab looks good:

```python
import os

import gitlab

GITLAB_TOKEN = os.getenv("GITLAB_TOKEN")
GITLAB_USER = os.getenv("GITLAB_USER")
GITLAB_URL = os.getenv("GITLAB_URL", "https://gitlab.com")
GITLAB_REPOSITORY = "3601513:gitterHQ/webapp"
GITLAB_DEFAULT_PAGE_SIZE = 100

repository_id = GITLAB_REPOSITORY.split(":")[0]

# private token or personal token authentication
client = gitlab.Gitlab(url=GITLAB_URL, private_token=GITLAB_TOKEN)
project = client.projects.get(id=repository_id)
requests = project.mergerequests.list(state="opened", order_by="created_at", sort="asc")

for request in requests:
    print(request.author["username"])
    print(request.created_at)
    print(request.merge_status)
    print(request.web_url)
    print(request.title)
    print(request.updated_at)
    print(" ")
```

**Example pull request reponse**

```json
{'_attrs': {'approvals_before_merge': None,
            'assignee': None,
            'assignees': [],
            'author': {'avatar_url': 'https://assets.gitlab-static.net/uploads/-/system/user/avatar/895869/avatar.png',
                       'id': 895869,
                       'name': 'Andrew Newdigate',
                       'state': 'active',
                       'username': 'andrewn',
                       'web_url': 'https://gitlab.com/andrewn'},
            'blocking_discussions_resolved': True,
            'closed_at': None,
            'closed_by': None,
            'created_at': '2017-07-04T11:48:27.863Z',
            'description': '\n'
                           '\n'
                           'Related to '
                           'https://gitlab.com/gitlab-org/gitter/webapp/merge_requests/1088',
            'discussion_locked': None,
            'downvotes': 0,
            'draft': False,
            'force_remove_source_branch': True,
            'has_conflicts': False,
            'id': 4109675,
            'iid': 1068,
            'labels': ['Category:Gitter',
                       'devops::create',
                       'group::gitter',
                       'section::dev'],
            'merge_commit_sha': None,
            'merge_status': 'cannot_be_merged_recheck',
            'merge_when_pipeline_succeeds': False,
            'merged_at': None,
            'merged_by': None,
            'milestone': None,
            'project_id': 3601513,
            'reference': '!1068',
            'references': {'full': 'gitterHQ/webapp!1068',
                           'relative': '!1068',
                           'short': '!1068'},
            'reviewers': [],
            'sha': '7c915e5e27bf6ded4eb0f7779bf22309d9bccc61',
            'should_remove_source_branch': None,
            'source_branch': 'feature/fix-anonymous-token-leak',
            'source_project_id': 3601513,
            'squash': True,
            'squash_commit_sha': None,
            'state': 'opened',
            'target_branch': 'develop',
            'target_project_id': 3601513,
            'task_completion_status': {'completed_count': 0, 'count': 0},
            'time_stats': {'human_time_estimate': None,
                           'human_total_time_spent': None,
                           'time_estimate': 0,
                           'total_time_spent': 0},
            'title': 'Fix anonymous token leak',
            'updated_at': '2020-09-02T17:20:10.829Z',
            'upvotes': 0,
            'user_notes_count': 3,
            'web_url': 'https://gitlab.com/gitterHQ/webapp/-/merge_requests/1068',
            'work_in_progress': False},
 '_module': <module 'gitlab.v4.objects.merge_requests' from '/Users/kyle/.pyenv/versions/3.9.4/lib/python3.9/site-packages/gitlab/v4/objects/merge_requests.py'>,
 '_parent_attrs': {'project_id': 3601513},
 '_updated_attrs': {},
 'approval_rules': <gitlab.v4.objects.merge_request_approvals.ProjectMergeRequestApprovalRuleManager object at 0x108bbf760>,
 'approvals': <gitlab.v4.objects.merge_request_approvals.ProjectMergeRequestApprovalManager object at 0x108bbfbe0>,
 'awardemojis': <gitlab.v4.objects.award_emojis.ProjectMergeRequestAwardEmojiManager object at 0x108bbf9d0>,
 'diffs': <gitlab.v4.objects.merge_requests.ProjectMergeRequestDiffManager object at 0x108bbf9a0>,
 'discussions': <gitlab.v4.objects.discussions.ProjectMergeRequestDiscussionManager object at 0x108bbfb80>,
 'manager': <gitlab.v4.objects.merge_requests.ProjectMergeRequestManager object at 0x108b8ccd0>,
 'notes': <gitlab.v4.objects.notes.ProjectMergeRequestNoteManager object at 0x108bbfc40>,
 'resourcelabelevents': <gitlab.v4.objects.events.ProjectMergeRequestResourceLabelEventManager object at 0x108bbfd00>,
 'resourcemilestoneevents': <gitlab.v4.objects.events.ProjectMergeRequestResourceMilestoneEventManager object at 0x108bbfcd0>,
 'resourcestateevents': <gitlab.v4.objects.events.ProjectMergeRequestResourceStateEventManager object at 0x108bbf8b0>}
```

**Comment by apoclyps** [comment]
a rough implementation has been implemented in https://github.com/apoclyps/reviews/pull/192

there are a number of todos around refactoring the existing codebase to make it reusable with other source control providers but it's looking promising. the only feature that is not strickly compatible is the lack of an addition/deletion diff on Gitlab merge requests.

**Comment by apoclyps** [comment]
#192 has been merged to master and will be available in release 3.0.0
